subsys/profiling/perf/perf.c

```diff
@@ -0,0 +1,98 @@
+/*
+ *  Copyright (c) 2023 KNS Group LLC (YADRO)
+ *  Copyright (c) 2020 Yonatan Goldschmidt <yon.goldschmidt@gmail.com>
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/arch/cpu.h>
+#include <zephyr/shell/shell.h>
+#include <zephyr/shell/shell_uart.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+size_t arch_perf_current_stack_trace(uintptr_t *buf, size_t size);
+
+struct perf_data_t {
+	struct k_timer timer;
+	int64_t end_time;
+
+	size_t idx;
+	uintptr_t buf[CONFIG_PROFILING_PERF_BUFFER_SIZE];
+};
+
+static struct perf_data_t perf_data = {
+	.end_time = 0,
+	.idx = 0
+};
+
+static void perf_tracer(struct k_timer *timer)
+{
+	size_t trace_length = 0;
+
+	if (++perf_data.idx < CONFIG_PROFILING_PERF_BUFFER_SIZE) {
+		trace_length = arch_perf_current_stack_trace(
+					perf_data.buf + perf_data.idx,
+					CONFIG_PROFILING_PERF_BUFFER_SIZE - perf_data.idx);
+	}
+
+	if (trace_length != 0) {
+		perf_data.buf[perf_data.idx - 1] = trace_length;
+		perf_data.idx += trace_length;
+	} else {
+		--perf_data.idx;
+		printf("perf buf override!\n");
```

Comment:
should print to the shell, have a look at [k_timer_user_data_set](https://docs.zephyrproject.org/apidoc/latest/group__timer__apis.html#gadba1884961e790dd9c5d567de91cc7e2)

Reply:
I had to create an additional handler thread because the shell_print does not work in interrupts.
I hope I understood correctly how you wanted me to use `k_timer_user_data_set`.

Reply:
Yeah I didn't know that the timer handler is called directly from an interrupt context, I managed to use a workqueue instead of a dedicated thread in https://github.com/zephyrproject-rtos/zephyr/commit/501c9093245fbc70df69ab3a07faa0055455e2dc (there are some other changes, as I was testing #73587)